### PR TITLE
Contravariant & Invariant

### PR DESCRIPTION
**Pull request by Cotel**
Adding Contravariant & Invariant typeclasses and instances.

### CODE REVIEW

File: modules/core/arrow-data/src/test/kotlin/arrow/data/MonoidTest.kt [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+@RunWith(KTestJUnitRunner::class)
+class MonoidTest : UnitSpec() {
+
+    val EQ: Eq<MonoidOf<Int>> = Eq.invoke { a, b ->
+        a.fix() == b.fix()
+    }
+
+    init {
+        testLaws(
+            InvariantLaws.laws(Monoid.invariant<Int>(), { Int.monoid() }, EQ)
```
Comment (Cotel) [inline modules/core/arrow-data/src/test/kotlin/arrow/data/MonoidTest.kt]: @pakoito It seems like our experiment success at compiling but tests are failing 🤔

File: modules/core/arrow-typeclasses/src/main/kotlin/arrow/typeclasses/Composed.kt [hunk 2]
```diff
@@ -150,6 +150,68 @@ fun <F, G> MonoidK<F>.compose(): MonoidK<Nested<F, G>> = object : ComposedMonoid
   override fun F(): MonoidK<F> = this@compose
 }
 
+interface ComposedInvariant<F, G> : Invariant<Nested<F, G>> {
+  fun F(): Invariant<F>
+
+  fun G(): Invariant<G>
+
+  override fun <A, B> Kind<Nested<F, G>, A>.imap(f: (A) -> B, g: (B) -> A): Kind<Nested<F, G>, B> =
+    F().run { unnest().imap({ ga -> G().run { ga.imap(f, g) } }, { gb -> G().run { gb.imap(g, f) } }) }.nest()
```
Comment (pakoito) [inline modules/core/arrow-typeclasses/src/main/kotlin/arrow/typeclasses/Composed.kt]: There's a chance this doesn't work on Android on the D8 compiler. Can you please move each of the lambdas in imap to a separate line?

File: modules/core/arrow-typeclasses/src/main/kotlin/arrow/typeclasses/Composed.kt [hunk 3]
```diff
@@ -192,7 +200,11 @@ interface ComposedInvariantContravariant<F, G> : Invariant<Nested<F, G>> {
     fun G(): Contravariant<G>
 
     override fun <A, B> Kind<Nested<F, G>, A>.imap(f: (A) -> B, g: (B) -> A): Kind<Nested<F, G>, B> =
-        F().run { unnest().imap({ ga -> G().run { ga.contramap(g) } }, { gb -> G().run { gb.contramap(f) } }) }.nest()
+        F().run {
+            val fl: (Kind<G, A>) -> Kind<G, B> = { ga -> G().run { ga.contramap(g) } }
+            val fr: (Kind<G, B>) -> Kind<G, A> = { gb -> G().run { gb.contramap(f) } }
+            unnest().imap(fl, fr)
```
Comment (pakoito) [inline modules/core/arrow-typeclasses/src/main/kotlin/arrow/typeclasses/Composed.kt]: Sorry, my bad. Take these out of the run block I mean.

Comment (Cotel) [reply modules/core/arrow-typeclasses/src/main/kotlin/arrow/typeclasses/Composed.kt]: 👌

**Comment by Cotel** [comment]
I've run all the tests locally and it seems like Monoid is passing Invariant Laws now!

So now I'm going to add an Invariant instance for Semigroup and that should be it for Invariant.

**Comment by pakoito** [comment]
>Can we fix this? 

Try making ForMonoid inherit ForSemigroup, as they're covariant. Subtyping sucks :) Test thoroughly.

**Comment by Cotel** [comment]
I'm also having some troubles to create a `Function1ContravariantInstance`. I've defined a `contramap` function in `Function1`:

```kotlin
class Function1<I, out O>(val f: (I) -> O) {
  fun <B> contramap(f: (B) -> I): Function1<B, O> = (this.f compose f).k()
}
```

But now I can't use it when I am defining the `ContravariantInstance`:

```kotlin
interface Function1ContravariantInstance<I> : Contravariant<Function1PartialOf<I>> {
  override fun <A, B> Kind<Function1PartialOf<I>, A>.contramap(f: (B) -> A): Function1<I, B> =
    fix().contramap(f)
}
```

The editor tells me that `contramap` will be called recursively. But even if I change it to `contramapp` to distinguish between scopes, the signature is incorrect. I would need it to be:

```kotlin
override fun <A, B> Kind<Function1PartialOf<I>, A>.contramap(f: (B) -> I): Function1<B, A>
```

**Comment by pakoito** [comment]
@Cotel Push it and I'll take a look

**Comment by pakoito** [comment]
@Cotel can't be done, same problem as the left functor from the other day

File: modules/core/arrow-data/src/test/kotlin/arrow/instances/ComposedInstancesTest.kt [hunk 4]
```diff
@@ -34,17 +35,34 @@ class ComposedInstancesTest : UnitSpec() {
...
+    val cf2: (Int) -> Kind<Nested<ForOption, Conested<ForFunction1, Int>>, Int> = { x: Int ->
+      Some({ y: Int -> x + y }.k().conest()).nest()
+    }
+
+    val bifunctorCf: (Int) -> Kind2<Nested<ForTuple2, ForTuple2>, Int, Int> = { Tuple2(Tuple2(it, it), Tuple2(it, it)).binest() }
 
     testLaws(
+      InvariantLaws.laws(ComposedInvariant(Option.functor(), NonEmptyList.functor()), cf, EQ_OPTION_NEL),
+      InvariantLaws.laws(ComposedInvariantCovariant(Option.functor(), NonEmptyList.functor()), cf, EQ_OPTION_NEL),
+      InvariantLaws.laws(ComposedInvariantContravariant(Option.functor(), Function1.contravariant<Int>()), cf2, EQ_OPTION_FN1),
```
Comment (pakoito) [inline modules/core/arrow-data/src/test/kotlin/arrow/instances/ComposedInstancesTest.kt]: TestLaws deduplicates tests based on name. That means only one of these will be run. Split 2 of them into a new call to testLaws.

**Comment by Cotel** [comment]
I've been looking for `Invariant` and `Contravariant` instances in cats but, besides the typeclasses stuff, these are all the instances we can make.

The only thing that remains is to test the laws for `ComposedContravariant`, `ComposedContravariantCovariant` and `ComposedCovariantContravariant` but these are really unintuitive for me, I am not able to create the required lambda for the law or the Eq instance 😢

**Comment by Cotel** [comment]
I would like to finish this asap and start something related with [comonadic UIs](https://docs.google.com/presentation/d/1UkRASOzbN67KinFq7SqJgsH_Q27iE4jxwgeveznWowM/edit#slide=id.g3d9244ddc6_0_361) but I was stuck and I'm quite busy right now. I feel bad for keeping this PR open for so long with no activity 😅

**Comment by pakoito** [comment]
I'm traveling, let's talk in September :D

**Comment by Cotel** [comment]
Hey! I might have some spare time tomorrow and I would like to close this and start working on another issue or something 😁

**Comment by Cotel** [comment]
I'm sorry but I really don't know what is failing here. I suppose that the laws which are failing are `ComposedInvariantContravariant` and the cause are probably the lambda and EQ instance for the test.

```kotlin
val cf2: (Int) -> Kind<Nested<ForOption, Conested<ForFunction1, Int>>, Int> = { x: Int ->
  Some({ y: Int -> x + y }.k().conest()).nest()                                          
}                                                                                        
```

```kotlin
val EQ_OPTION_FN1: Eq<NestedType<ForOption, Conested<ForFunction1, Int>, Int>> = Eq { a, b ->
  a.unnest().fix() == b.unnest().fix()                                                       
}                                                                                            
```

Any ideas? 😅

**Comment by pakoito** [comment]
I may take a look directly in the afternoon. Meanwhile, use the extfun `Eq.logged` and run the tests with `-i` to at least see the values.

**Comment by Cotel** [comment]
God bless that `Eq.logged` function 😍.

**Comment by pakoito** [comment]
@Cotel It's saved my ass more times that I'm willing to admit hahaha

Gotta give a last check and we'll go green ;)

File: modules/core/arrow-data/src/test/kotlin/arrow/instances/ComposedInstancesTest.kt [hunk 5]
```diff
@@ -36,7 +36,15 @@ class ComposedInstancesTest : UnitSpec() {
       }
 
     val EQ_OPTION_FN1: Eq<NestedType<ForOption, Conested<ForFunction1, Int>, Int>> = Eq { a, b ->
-      a.unnest().fix() == b.unnest().fix()
+      a.unnest().fix().fold(
```
Comment (pakoito) [inline modules/core/arrow-data/src/test/kotlin/arrow/instances/ComposedInstancesTest.kt]: Jesus fucking christ

File: modules/core/arrow-data/src/test/kotlin/arrow/data/ConstTest.kt [hunk 6]
```diff
@@ -16,6 +13,7 @@ class ConstTest : UnitSpec() {
   init {
     ForConst(Int.monoid()) extensions {
       testLaws(
+        InvariantLaws.laws(this, { Const(it) }, Eq.any()),
```
Comment (pakoito) [inline modules/core/arrow-data/src/test/kotlin/arrow/data/ConstTest.kt]: InvariantLaws should be part of FunctorLaws, so it's tested by ApplicativeLaws!

File: modules/core/arrow-data/src/test/kotlin/arrow/instances/ComposedInstancesTest.kt [hunk 7]
```diff
@@ -34,17 +35,48 @@ class ComposedInstancesTest : UnitSpec() {
...
+    testLaws(
+      InvariantLaws.laws(ComposedInvariantCovariant(Option.functor(), NonEmptyList.functor()), cf, EQ_OPTION_NEL)
+    )
+
+    testLaws(
+      InvariantLaws.laws(ComposedInvariantContravariant(Option.functor(), Function1.contravariant<Int>()), cf2, EQ_OPTION_FN1)
+    )
 
     testLaws(
+      InvariantLaws.laws(ComposedInvariant(Option.functor(), NonEmptyList.functor()), cf, EQ_OPTION_NEL),
```
Comment (pakoito) [inline modules/core/arrow-data/src/test/kotlin/arrow/instances/ComposedInstancesTest.kt]: Same as above, InvariantLaws should be part of FunctorLaws.

File: modules/core/arrow-test/src/main/kotlin/arrow/test/laws/ContravariantLaws.kt [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+import arrow.test.generators.genFunctionAToB
+import arrow.typeclasses.Contravariant
+import arrow.typeclasses.Eq
+import io.kotlintest.properties.Gen
+import io.kotlintest.properties.forAll
+
+object ContravariantLaws {
+
+    inline fun <F> laws(CF: Contravariant<F>, noinline cf: (Int) -> Kind<F, Int>, EQ: Eq<Kind<F, Int>>): List<Law> =
+        listOf(
```
Comment (pakoito) [inline modules/core/arrow-test/src/main/kotlin/arrow/test/laws/ContravariantLaws.kt]: In here ContravariantLaws imply InvariantLaws also pass, so replace this line with:

InvariantLaws.laws(...) + listOf(

File: modules/core/arrow-typeclasses/src/main/kotlin/arrow/typeclasses/Composed.kt [hunk 9]
```diff
@@ -211,20 +348,20 @@ interface ComposedBifoldable<F, G> : Bifoldable<Nested<F, G>> {
 
   fun G(): Bifoldable<G>
 
-  override fun <A, B, C> BiunnestedType<F, G, A, B>.bifoldLeft(c: C, f: (C, A) -> C, g: (C, B) -> C): C = F().run {
+  override fun <A, B, C> BinestedType<F, G, A, B>.bifoldLeft(c: C, f: (C, A) -> C, g: (C, B) -> C): C = F().run {
```
Comment (pakoito) [inline modules/core/arrow-typeclasses/src/main/kotlin/arrow/typeclasses/Composed.kt]: Are we sure of this change, or was it part of the merge?

Comment (Cotel) [reply modules/core/arrow-typeclasses/src/main/kotlin/arrow/typeclasses/Composed.kt]: I've checked the history and this is part from our changes, related to https://github.com/arrow-kt/arrow/pull/975/commits/fa50cf1debda58318b1ae59d3666a10d1aeca685

Comment (pakoito) [reply modules/core/arrow-typeclasses/src/main/kotlin/arrow/typeclasses/Composed.kt]: Right, right, they're correct, there's a couple of fixes to other that were wrong.

Cool, only the test left then.

File: modules/core/arrow-data/src/test/kotlin/arrow/instances/ComposedInstancesTest.kt (hunk 7 above)
Comment (Cotel) [reply modules/core/arrow-data/src/test/kotlin/arrow/instances/ComposedInstancesTest.kt]: FuntorLaws are being used at line 74. Should I remove that line too?

**Comment by pakoito** [comment]
🎉

**Pull request merged**